driver-core/src/main/com/mongodb/client/model/expressions/Expression.java

```diff
@@ -21,116 +21,200 @@
+     */
     IntegerExpression isIntegerOr(IntegerExpression other);
+
+    /**
+     * {@code this} value as a {@link BooleanExpression boolean} if
+     * {@code this} is a boolean, or the {@code other} boolean value if
+     * {@code this} is null, or is missing, or is of any other non-boolean type.
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     */
     StringExpression isStringOr(StringExpression other);
+
+    /**
+     * {@code this} value as a {@link StringExpression string} if
+     * {@code this} is a string, or the {@code other} string value if
+     * {@code this} is null, or is missing, or is of any other non-string type.
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     */
     DateExpression isDateOr(DateExpression other);
+
+
+    /**
+     * {@code this} value as a {@link ArrayExpression array} if
+     * {@code this} is an array, or the {@code other} array value if
+     * {@code this} is null, or is missing, or is of any other non-array type.
+     *
+     * <p>Warning: this operation does not guarantee type safety. While this
+     * operation is guaranteed to produce an array, the type of the elements of
+     * that array are not guaranteed by the API. The specification of a type by
+     * the user is an unchecked assertion that all elements are of that type,
+     * and that no element is null, is missing, or is of some other type. If the
+     * user cannot make such an assertion, some appropriate super-type should be
+     * chosen, and elements should be individually type-checked.
+     *
+     * @param other the other value.
+     * @return the resulting value.
+     * @param <T> the type of the elements of the resulting array.
+     */
     <T extends Expression> ArrayExpression<T> isArrayOr(ArrayExpression<T> other);
+
+    // TODO-END doc after Map merged, "record" and "schema objects" are decided
     <T extends DocumentExpression> T isDocumentOr(T other);
 
 
     /**
-     * server error if type cannot be converted to string (arrays, objects)
+     * The {@link StringExpression string} value corresponding to this value.
      *
-     * @return
+     * <p>This may "cause an error" if the type cannot be converted to string,
+     * as is the case with {@link ArrayExpression arrays},
+     * {@link DocumentExpression documents}, and {@link MapExpression maps}.
+     * TODO-END what about null/missing?
+     * TODO-END document vs record
+     * TODO-END "cause an error" above
+     *
+     *
+     * @see StringExpression#parseDate()
+     * @see StringExpression#parseInteger()
+     * TODO-END all the others? implement?
+     * @return the resulting value.
      */
     StringExpression asString();
 
     /**
-     * Applies the given function to this argument. Note that "apply" usually
-     * applies functions to arguments; here, the parameters are reversed.
+     * Applies the provided function to {@code this}.
```

Comment:
If we are making it vague whether `isBooleanOr` (and similar methods) return `this`/`other` references or not, then we probably should use that approach here and say `Applies the provided function to {@code this} value.`

Note that the documentation of `Expression.switchMap` says _`applying the provided switch mapping to {@code this} value`_, not `... to {@code this}`, which is consistent with that I am proposing here.

Reply:
I agree.